### ISSUE TITLE
Model forecasting

### ISSUE THREAD

**Issue by sylvainschmitt** [sylvainschmitt/SSDM#42]
@lukasbaumbach, this the new issue for the development of model forecasting within `SSDM`without the trick from #10 .

**Comment by lukasbaumbach** [comment]
@sylvainschmitt I think we can make this work by using your project function on the lowest level,  feeding it with the decomposed single models stored in the ESDMs and SSDMs and putting everything back together afterwards. 

However, there seems to be a missing link (as also mentioned by a user in issue #10) : as far as I could see there is no enms slot containing the single SDMs in the Ensemble.SDMs produced by ensemble_modelling. Should this slot exist by default? There was no error and the ESDM looks fine.

**Comment by sylvainschmitt** [comment]
@lukasbaumbach effectively the `ensemble` and `sum` methods used to build an ESDM from several SDMs do not keep the SDMs in the ESDM object (they are lost in the process), contrarily to ESDMs that are kept within the SSDM. So we might discuss, if needed for model forecasting, how to keep SDMs in the ESDM (and thus even in the final SSDM), and more particularly **what** to keep. Because there is also a question of object size and memory use that is a constant issue with species distribution modelling. And we need to think if we want to keep each full SDM object within the ESDM or only part of the information.

**Comment by sylvainschmitt** [comment]
@lukasbaumbach just one thing, you suggest here to change a little the architecture of the package by adding a slot to the ESDM object, which is okay and certainly necessary to add model forecasting to the package. But just for you to know, regarding discussion about architecture changes in the package I will need to add co-atuhors in the discussion / decisions.

**Comment by lukasbaumbach** [comment]
@sylvainschmitt actually I thought the slot would already exist, but in that case - yes we would need it in order to project the models. In terms of memory use, there could be issues with a large number of replicates. In that case, the package could offer the possibility to write the models separately during ensembling, such that the user can access them later for projection.
But yeah, these are only suggestions and the authors should decide, what should happen with the package :)

**Comment by sylvainschmitt** [comment]
@lukasbaumbach the slot was not kept as we judged it useless for the end use of the ESDM, but for model forecasting it seems mandatory. Writing raster and temporary files is already implemented to free the RAM (but not in that specific case), nevertheless it might produce a lot of data in the temporary files. I think the question is also about if we want (1) model forecasting to be available any time with any new environments from an existing SDM/ESDM/SSDM or if we want (2) model forecasting to be chosen when launching the model building with a second environmental dataset.

**Comment by lukasbaumbach** [comment]
Since most climate impacts studies try to (and should) consider several climate scenarios when projecting models, I guess it would be favourable to make the projection possible anytime after creating the model. Also including a second projection into model building just to realize that the model still needs some tweaking can be quite frustrating (speaking from experience).

Yet given the fact, that different users have different demands and preferences, I would suggest another approach. What bothered me a little about the package was that it was not possible to build replicate models with the modelling() function. Why not include the possibility to build a list of models (maybe even of different algorithms), which is not ensembled yet? Memory may become an issue with many replicates, but this is hardly avoidable anyways. 
This list could then be the basis for further steps such as using the project() function on them, putting them together with ensemble() and stacking the ensembles with stacking(). 

The only trouble in this process would be to do this for several species simultaneously, but maybe you got an idea?

**Comment by sylvainschmitt** [comment]
Yes I think you are right, besides being more complicated, allowing to reproject a model already built any time would be the best behavior for the package.

Let's not think directly about memory right now, and deal with it once we see what we want to implement. I am not quite sure to have understood what you meant by building a list of model. You mean be able to launch several modelling bracketing some parameters ? I think it's a different issue but can be easily implemented. Could you give a dummy example of what you would expect to be able to do in a different issue (maybe related to this one after) ?

I think doing it for several species is no issue at all, it's always the same process repeated on several objects, it's just taking a lot of resources (memory, CPU and time). And we should keep in mind tat the package was aimed at practitioners too, so it should be still able to run relatively easily on classical laptop. And a last point that is time consuming that we should think of is to keep the Graphical User Interface (GUI) alive. for the moment the GUI includes all the functions (but it's not mandatory, especially in first developments). But it should at least always be able to run basic SSDM modelling. And the bade side of the GUI, it that it does not include formal testing. So we have to check it manually from time to time to be sure to have it running for end-users.

**Comment by lukasbaumbach** [comment]
Sorry for getting back to this rather late, I shifted my workplace to Costa Rica for the next month.
So, you are right with closing the issue, the replications can already be accomplished with the ESDM. The only necessary addition then is to include ESDM@SDM slots in the ESDM. Should we just go ahead and try that? I'm not very familiar with the creation of S4-classes, but with a little help from your side I can try (or if it's a no-brainer for you, just go ahead).

**Comment by sylvainschmitt** [comment]
No worry, I'm happy you're still motivated to extend `SSDM` and we are in no urge as it wouldn't happen in a near future without you. I think you could start to try on a different development branch to add an `ESDM@SDM` slot and test if computation, memory, and everything is still ok for basic small examples (included in the package) and bigger data-sets. In order to do that, you'll need to learn/follow two things:

1. The development guidelines (which might be more or less general or specific to me), including code writing, documenting, formal testing, checking, tracking and sharing. Unfortunately, I kind of evolved as any developers and changed a bit my way of coding, so you can see things coded that I would not advise (and you are free to update them while you don't break the code). Anyway, in another project I wrote some guidelines that you can have a look to (https://sylvainschmitt.github.io/RconTroll/developers_guidelines.html) and generally I try to follow Hadley Wickham's book on R package recommendations (http://r-pkgs.had.co.nz/)
2. Object-Oriented programming in R, which is not very known but very useful. Because basically, SDM, ESDM, and SSDM are R objects with attributes (slot) and methods. Have a look to Hadley Wickham's book on Advanced R (yes him again, http://adv-r.had.co.nz/OO-essentials.html).

Maybe I would go quicker by myself but I'm lacking of time, and it's better you start to dive into the code the sooner, especially regarding S4, to be independent the quicker. But don't worry I can guide you into the architecture at the beginning to avoid a lost of time trying to understand the architecture without help.

**Comment by sylvainschmitt** [comment]
Globally S4 class start with a capital and have a comma in their name. You'll find ESDM called `Ensemble.SDM` definition here https://github.com/sylvainschmitt/SSDM/blob/master/R/Ensemble.SDM.R. And the `ensemble` and `sum` methods for `Algortihm.SDM` class are the methods that allow to transform several SDMs in one ESDM (so this is where you need to deal with the filling of your new slot called SDM), it is available here https://github.com/sylvainschmitt/SSDM/blob/master/R/ensemble_modelling.R. Finally, the `ensemble` method is called in the `ensemble_modelling` and thus `stack_modelling` function, where you need to add parameter if you add some to the `ensemble` method (which shouldn't be the case here).

**Do not hesitate to ask for further information.**

**Comment by RobinPouteau** [comment]
Hi Lukas, it has been a while since Sylvain and I talked about implementing a forecasting function in SSDM and I was very happy to hear that you are keen to tackle the challenge. In addition to project distributions into the past or future, I would like you to keep in mind that one could also want to use this function to project a model at a different resolution (down- or up-scaling).

**Comment by sylvainschmitt** [comment]
@RobinPouteau I'm not very clear on what you wish the user to have access to regarding projection at different resolutions. There is two different cases for me. Regarding up-scaling, it would mean that the user have environmental descriptors with increased resolution of the area, so why not use them directly in the first modelling to be more accurate ? Regarding down-scaling, why not built a model at the highest resolution and down-scale prediction using the `aggregate` method for rasters in the `raster` package ? Anyway, if the model forecasting method is just a projection with a different environmental dataset including the same variables as the original dataset, it should do what you wish, *e.g.*:
```
DistributionModelAtDifferentResolution <- forecast(DistributionModel, EnvironmentAtDifferentResolution)
```

**Comment by lukasbaumbach** [comment]
> Regarding down-scaling, why not built a model at the highest resolution and down-scale prediction using the aggregate method for rasters in the raster package ?
@sylvainschmitt I think you are actually confusing up-scaling with down-scaling here. Down-scaling implies going from a low to a higher resolution, hence generating information at a smaller scale than before. Anyways, Sylvain is right, any projection to a different resolution can just be done with the forecast function, since it only applies any given model to a new environment regardless of the  resolution of these environmental rasters.

Since my project is steering towards the SDM phase again, I'll try to busy myself with package development over the next couple weeks. I'll keep you posted about any progress :)

**Comment by lukasbaumbach** [comment]
Hi,
so I managed to insert a slot (for now it's a list called `ESDM@individual.models`) into the ESDM class, tested it and everything still seems to work (at least `modelling`, `ensemble_modelling`, `ensemble` and `project`) ! I also started tweaking the `project` function, so it can take ESDM objects directly (using `lapply`). Furthermore, I tested memory consumption with the example code using `mem_change` (pryr). The difference was small for an ESDM, but it really bloated for an SSDM (now twice the size). This brings me to a few questions and suggestions as to how to proceed:

1. It would be nice to give the users a little more flexibility in choosing only the outputs they need (i.e. insert function arguments like "individual.models=TRUE/FALSE","projection.only=TRUE/FALSE", you get the idea)
2. This applies especially for the projections, since there is no real additional value in returning the original models alongside the rasters. By default returning a raster for `Algorithm.SDM`; a list/stack of rasters and/or mean raster for `Ensemble.SDM` and a list of lists and/or biodiversity map (sum of binarized projections) and/or mean of all projections for `Stacked.SDM` might be a simpler option
3. NOTE: The original ensemble function only keeps models that meet the `ensemble.metric` and `ensemble.thresh` criteria. I did not manage to only store this selection in the slot yet, but I can try again later.
4. NOTE: I'll have another test of memory and time consumption with my data soon.
5. My next step would be to further modify the project function, so it can take anything (SDM/ESDM/SSDM) and externalize it, so it can be used directly (without calling SSDM:::project)

I'm curios on your opinions!

**Comment by sylvainschmitt** [comment]
Hi,

Congrats, it seems that you worked well. But before answering I wanted to set few things. Where are you developing and how ? Because the new code should comply with SSDM developing habits, including implementing formal testing and arguments test when developing new code. The best would be for you to develop on a new branch forked from my GitHub repository. Especially because it will help you to catch up with parallel development made in SSDM. For instance, I recently had an issue with a change in raster that brought a dependency issue and a bug found in the load function, that were fixed in the new version. 

Concerning what you did. It's nice that you opened this new slot, but we should agree together on the objects names. Because SDM field can be a bit complicated we tried in the article to have a formal nomenclature. Besides it's not respected as the SSDM slot is called enms instead of ESDMs. But I think ESDM@SDMs would be better than ESDM@individuals.models (especially because the . is normally dedicated to methods). If you want to know if everything works you can use the formal testing implemented in the package with devtools::test() or CTRL+SCHIFT+T in Rstudio. Furthermore, it's also nice that you started tweaking `project`, but as you may have seen it's not a function but a method. So implementing it for class `Ensemble.SDM` is nice, if it's done by the rule. Regarding memory it seems logical as an SSDM is quickly accumulating every overcharge. So before addressing your questions, try to have a look to mine.

Concerning your questions:

1. I totally agree that if we have memory issue the user should be able to tweak outputs to adapt to his needs. Anyway we should try not to over extend the number of parameters per function, and any choice of default value should be discussed (the idea being that the next version default behavior stay similar to the current). Another thing, when you change classes or method we always need to think to backward compatibility. If a new slot is opened or a new method is created. A user with a saved object of an older version of SSDM should be able to work with the new version of SSDM. So we should start discussing the new options (for instance I'm for individual.models or SDMs with default FALSE in `ensemble` and `ensemble.model` but not projection.only).
1. Yes we could consider a kind of "minimal outputs" option, trying not to break any functions or methods and being compatible with all the pipeline from SDM to SSDM.
1. Keeping only the selected SDMs in the ESDM seems mandatory. Because I don't know how you did it I can't help you without the code (cf my first comment).
1. OK
1. I think we should first open properly the `project` method to the user documenting it before extending it to new classes. But glovbally I agree to extend it to Ensemble.SDM and Stack.SDM for forecasting purposes.

I hope my opinions didn't disappointed you. I'll try to start an objective list as a base of discussion in the ISSUE to help organize the discussion. And at a certain point we could schedule a Skype if needed to help you progress.

**Comment by sylvainschmitt** [comment]
## #42 implementations

*All implementation should include documentation, automatic argument checking (`.checkargs`) and formal testing (test folder) before being closed.*

- [ ] Open a new branch and use social coding on GitHub for new implementations
- [ ] Implement of ESDM@SDMs slot
    - [ ] Chose the correct name
    - [ ] Keep only the selected SDMs
- [ ] Develop `project` method
    - [ ] Open it to the user and document it
    - [ ] Implement projection with new environment
    - [ ] Extend the method to the `Ensemble.SDM` class
    - [ ] Extend the method to the `Stack.SDM` class
- [ ] Develop new options for `ensemble`, `ensemble_modelling`, `stacking`, `stack_modelling`
    - [ ] Chose and discuss options
    - [ ] `SDMs = FALSE` as a first one ?
    - [ ] `minimal.outputs = FALSE` ? 

*There is no specific order in the list, and the post is made to be discussed and changed.*

**Comment by lukasbaumbach** [comment]
Hi, I'm not disappointed whatsoever. This was just a test I quickly did on my local machine (using the latest Git version of the package) and the formal coding and documentation will follow. But yes, I will make everything traceable on Git from now on.
It's good that you mentioned the slot naming convention, since this is something I don't know (I could not really derive anything from 'enms'). Again, it was only provisional. Maybe 'sdms' (lower case) would fit?
If it's okay for you, I'll try working down the list on a new branch. And sure, we can set up a Skype when I got a little further.

**Comment by sylvainschmitt** [comment]
Yes I think `sdms` in lower case should be best. And `enms` should be replaced by `esdms` in the same time (ENM stand of r Ensemble Niche Model, but there is dozens of ways to call SDMs and I think we should follow the article names). Anyway, every general convention regarding the package should be approved by coauthor. Maybe when will have a list of objects we can discuss them together. Good luck for the next steps.

**Comment by sylvainschmitt** [comment]
@lukasbaumbach , following your Pull Request, I add the ToDo list to this issue (to be updated):
- [x] `project` methods description
- [x] `project` methods  test units
- [x] `project` methods  input check
- [ ] update GUI
- [ ] vignette about model forecasting
- [ ] CRAN submission (maybe after other implementations) ?

**Comment by lukasbaumbach** [comment]
@sylvainschmitt sounds good! 
As a further implementation I guess one would also need to think of parallelization of `project.`  At the same time, the parallelization techniques of `ensemble_modelling` and `stack_modelling` could be overhauled, e.g. by using the `future` package as suggested in #13 .

**Comment by sylvainschmitt** [comment]
Parallelization is a more complex issue, because you can always discuss of the level were you want to parallelize ! We made the choice to parallelize using `parallel` (and not `future` because I did not knew it at that time) at the SDM level (kind of a high level). But definitely `project.Esemble.SDM` and `project.Stack.SDM` could use the same parallelization procedure as `ensemble_modelling` and ` stack_modelling`.

**Comment by lukasbaumbach** [comment]
Yes, the parallelization options are already applied at the right places (i.e. whenever an ESDM or SSDM is created). 

I was just thinking, also following your positive comment about implementing `future` in the next version of SSDM (see issue 13), that if parallelization is to be newly included for `project`, why not update it with the latest techniques? Effectively, as discussed in issue 13, basically only `parLapply` has to be replaced with future_lapply (and you don't need the `clusterExport` call anymore). 
I just recently tried `future_lapply` for a code example and it was incredibly fast.
So if you're willing to try, we could give it a shot.

**Comment by sylvainschmitt** [comment]
You make me curious ! Actually I never used `future_lapply` (so I did a lot of `clusterExport`). Are you sure it deals well is the environments ? Because `clusterExport` might seem a bit unfriendly, but it allows to have a really light exported environment. But okay you can give it a shot.

**Comment by sylvainschmitt** [comment]
Just one very important thing that is written anywhere, you may have noticed that parallel computing is called even when asking for only one core:
https://github.com/sylvainschmitt/SSDM/blob/f10c2c9302941e5abba0238032b0642d71ebe318/R/stack_modelling.R#L335
which might seems strange, as 1 core should be serial computing. But there is a reason for that. At the end of the development of `SSDM` we tested its robustness on HPC. And we noticed that there was few memory leak issues. They can seem harmless when testing small models, but as soon as your start huge SSDM they can be very painful. One of the main memory leak (that might have been solved by know) was related to the java session called by `MAXENT`. So because we couldn't track and fix the memory leak (it's inside a dependency). We used this work-around of always doing parallel computing even on one node. Like that, R open a new environment for the ESDM creation and kills it after, resulting in memory leak to be removed after each ESDM. So it can seems a little detail, but it has huge consequences when up-scaling analyses. Consequently we should take care it stays like that.

**Comment by lukasbaumbach** [comment]
Hi,
I've been playing around with the GUI and first of all have to say: it's no fun! Anyways, I managed to create an output table, which shows the selected environmental variables. Also I created a tab for forecasting and almost got it to load the updated rasters (using the new data tab code). However, I reached a point where every further change seems to result in an error, so I'd be happy if you can explain me what I might be doing wrong. 

I created new slots in `input` and `data` (i.e. updated the names from e.g. `input$envfiles` to `input$projenvfiles`) and inserted a function to compare the names of the old raster layers with the new/updated ones (later I also wanted to have a warning, if they don't match). When loading, it seems like only one layer is loaded (I get errors like `data$ProjEnv` not behaving like a raster stack and can't change the preview in the dropdown menu).

Is it possible that the `load.var` function needs to be changed, too? It's hard for me to keep an overview over the `server` functions, since I'm still new to shiny. What do I need to do to apply the same functions (like load and plot rasters) to different data?

We can also open a new thread (e.g. pull request with the current edits), if we're getting too much off course of the forecasting topic.

**Comment by sylvainschmitt** [comment]
Hi @lukasbaumbach ,

> I've been playing around with the GUI and first of all have to say: it's no fun! 

Yes the use of shiny can be a bit complicated, and to add complexity the whole GUI is only one UI file, so a bit hard to get it. Worst, without diving into the code it's a bit hard for me to follow you. But I think the error is here:

> I created new slots in `input` and `data` (i.e. updated the names from e.g. input$envfiles to input$projenvfiles)

You can create a reactive slot in `data` called `projenvfiles` for additional environmental files. But you can"t do the same with the `input` object as the `input` correspond to the UI objects in `ui.R`. For instance `input$envfiles` is automatically created (not by me in the server) by the following code in the UI:
https://github.com/sylvainschmitt/SSDM/blob/80bba1b61cb3dcfa7c7f25f1cb33555613d09dae/inst/shiny/gui/ui.R#L32

Anyway you should never change the basic R function as `load.var`. 

And **most important** how do you plan to add model forecasting to the GUI ? Because it seems to me that you are trying to replace `input$envfiles` by `input$projenvfiles` for model forecasting. Which you shouldn't ! The GUI should function has before and integrate the novelty. Do you plan to include forecasting when the user is starting a new model or with an already existing model ? I think we should clarify the behavior of the GUI before implementing it. Anyway I could help you develop a bit the GUI if needed.

**Comment by lukasbaumbach** [comment]
I was still at the stage of testing and trying to replicate the new data page for loading projected layers.
But yes, I already had a few design questions in mind, which I was going to ask you:
1) Should it be possible to only update a few layers (e.g. climatic) and leave the others as is? Or should the user again provide all layers? The new environmental data could then be loaded into `data$ProjEnv`.
2) Can you briefly describe how the environmental data is loaded/stored when running the application? Which parts of the data loading code need to be adjusted for loading new data and which don't? 
3) I'm also trying to implement a warning, if the layer names don't match the variables used for model training. Should I do this like you did for `Envbug`?
4) In my opinion it would be preferable to include the step of projection AFTER creating the model. Otherwise the modelling might become overkill and also much more complicated to debug.

So basically I would add a new tab "Model forecasting" which appears after building a model. The page would look like the new data tab (Raster selection, categorical variables and load button, uiOutput for error messages, Preview on the right) with an added "project" or "forecast" (a common terminology would be good at this point) button at the bottom of the page. Upon completion, a new "results" page could appear, showing the most important results (simple map for SDM, ensemble and uncertainty map for ESDM and diversity, endemism and uncertainty map for SSDM).

By the way: I dropped the idea of trying to implement parallelization with the `future` package again. It seems due to instability issues some of the techniques don't work for RStudio anymore, which I reckon is one of the most common platforms for running R. Let's just try and implement `parLapply` for project then :)

**Comment by sylvainschmitt** [comment]
> Should it be possible to only update a few layers (e.g. climatic) and leave the others as is? Or should the user again provide all layers? The new environmental data could then be loaded into data$ProjEnv.

Regarding the GUI it always depend on how much time and effort you want to spend on it. For the user it will be nicer to only update a few layers. But for you it will mean much more work, whereas the user would only have to reload unchanged variables. So do as you want, but keep in mind that we invest a lot of robustness in the code, but we do not seek for the perfect GUI as we are R developer and not software maker.

**Comment by sylvainschmitt** [comment]
> Can you briefly describe how the environmental data is loaded/stored when running the application? Which parts of the data loading code need to be adjusted for loading new data and which don't?

Briefly no ! But let's go:

Environmental variable are loaded in the following code and stock in the variable `input$envfiles` thanks to the `shinyFilesButton()`
https://github.com/sylvainschmitt/SSDM/blob/80bba1b61cb3dcfa7c7f25f1cb33555613d09dae/inst/shiny/gui/ui.R#L30-L36

Then, data in the server are initialized with the `data` reactive values here
https://github.com/sylvainschmitt/SSDM/blob/80bba1b61cb3dcfa7c7f25f1cb33555613d09dae/inst/shiny/gui/server.R#L13

And similarly variable information are stocked in the `load.var ` reactive values:
https://github.com/sylvainschmitt/SSDM/blob/80bba1b61cb3dcfa7c7f25f1cb33555613d09dae/inst/shiny/gui/server.R#L51

Than files are selected with the `shinyFileChoose` tool on the server side, defining thus `input$envfiles`:
https://github.com/sylvainschmitt/SSDM/blob/80bba1b61cb3dcfa7c7f25f1cb33555613d09dae/inst/shiny/gui/server.R#L68-L79

And finally if everything is good, environmental variables are loaded in the `data$Env` slot when pressing the `input$load` shinyButton:
https://github.com/sylvainschmitt/SSDM/blob/80bba1b61cb3dcfa7c7f25f1cb33555613d09dae/inst/shiny/gui/server.R#L90-L138

I'm still open to question, as shiny ui/server logic can be a bit hard. And I can help you to write a bit of code at this stage.

**Comment by sylvainschmitt** [comment]
> I'm also trying to implement a warning, if the layer names don't match the variables used for model training. Should I do this like you did for Envbug?

Yes totally.

> In my opinion it would be preferable to include the step of projection AFTER creating the model. Otherwise the modelling might become overkill and also much more complicated to debug.

Yes and no. First don't forget that instead of creating a model you can also load a previous one with the GUI. I think I would do it another way. Let's see with next comment.

> So basically I would add a new tab "Model forecasting" which appears after building a model. The page would look like the new data tab (Raster selection, categorical variables and load button, uiOutput for error messages, Preview on the right) with an added "project" or "forecast" (a common terminology would be good at this point) button at the bottom of the page. Upon completion, a new "results" page could appear, showing the most important results (simple map for SDM, ensemble and uncertainty map for ESDM and diversity, endemism and uncertainty map for SSDM).

I'm not sure about this. First I would definitely not add a new result page, it would mean duplicate a lot of existing code and overcomplexifying the GUI. Instead, why not update the result page with the forecasted model ? In fact, if you only replace the `data$ENM` or `data$Stack` with the updated model, the page should update itself (magic no ?) ! Similarly, the new page you described for "model forecasting" tab seems a lot of repetition of the load tab. Instead I would reuse the "load" tab to update the rasters (with maybe a raster update button), and I would add a forecast option in the Modelling type selection of the modelling tab. But it's not a definitive choice. I think we might discuss it with @RobinPouteau , @dimitri-justeau and @floriandeboissieu .

**Comment by lukasbaumbach** [comment]
Okay, thank your for the guidance! In summary:
- don't create a new page for projection but add the possibility to update the `data$Env` in the new data page
- let the user pick all variables for projection, make sure layer names match training variable's names
- add a "projection" button (probably as 4th option inside the modelling dropdown menu)
- take the user directly to the results page, once projection is done (could use reactive map title to make it clear, that the results are updated)
> the page should update itself (magic no ?)

haha true!

> And I can help you to write a bit of code at this stage.

Thanks, I'll get back to that for sure!

**Comment by lukasbaumbach** [comment]
The new version is nearing completion! Just a formal question: should we rename the option whether to update the projection slots or return a raster list in the project function? Currently it's called "update.projections" (TRUE/FALSE). Would something like "output.format" ('model'/'rasters') be more intuitive?  Or keep the current name?

**Comment by sylvainschmitt** [comment]
Great news ! output.format seems clearer to me. But both are okay

**Comment by lukasbaumbach** [comment]
implemented in newest version
